Description: Don't perform chown if target already has the expected ownership.
Forwarded: not-needed
Origin: self
Author: Frode Nordahl <frode.nordahl@canonical.com>

index 499862337..df0654d70 100644
--- a/src/ceph-volume/ceph_volume/util/system.py
+++ b/src/ceph-volume/ceph_volume/util/system.py
@@ -1,4 +1,5 @@
 import errno
+import grp
 import logging
 import os
 import pwd
@@ -108,7 +109,9 @@ def mkdir_p(path, chown=True):
             raise
     if chown:
         uid, gid = get_ceph_user_ids()
-        os.chown(path, uid, gid)
+        path_stat = os.stat(path)
+        if path_stat.st_uid != uid or path_stat.st_gid != gid:
+            os.chown(path, uid, gid)
 
 
 def chown(path, recursive=True):
@@ -116,6 +119,16 @@ def chown(path, recursive=True):
     ``chown`` a path to the ceph user (uid and guid fetched at runtime)
     """
     uid, gid = get_ceph_user_ids()
+    path_stat = os.stat(path)
+    # NOTE: This function may be used to change ownership of the LVM VG/LV
+    # symlinks under /dev. On Ubuntu these are owned by the group 'disk'.
+    try:
+        disk_gid = grp.getgrnam('disk').gr_gid
+    except KeyError:
+        disk_gid = 0
+    if path_stat.st_uid == uid and (
+            path_stat.st_gid == gid or path_stat.st_gid == disk_gid):
+        return
     if os.path.islink(path):
         process.run(['chown', '-h', 'ceph:ceph', path])
         path = os.path.realpath(path)
